modules/shredder/src/main/scala/com/snowplowanalytics/snowplow/shredder/Discovery.scala

```diff
@@ -0,0 +1,120 @@
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.shredder
+
+import cats.syntax.either._
+
+import com.amazonaws.{AmazonClientException, AmazonWebServiceRequest, ClientConfiguration}
+import com.amazonaws.retry.{PredefinedBackoffStrategies, RetryPolicy}
+import com.amazonaws.services.s3.{AmazonS3, AmazonS3ClientBuilder}
+import com.amazonaws.services.sqs.{AmazonSQSClientBuilder, AmazonSQS}
+import com.amazonaws.services.sqs.model.SendMessageRequest
+
+import com.snowplowanalytics.iglu.core.circe.implicits._
+
+import com.snowplowanalytics.snowplow.rdbloader.generated.ProjectMetadata
+import com.snowplowanalytics.snowplow.rdbloader.common.{S3, Cloud, LoaderMessage, Semver}
+import com.snowplowanalytics.snowplow.rdbloader.common.S3.Folder
+
+
+object Discovery {
+
+  final val FinalKeyName = "shredding_complete.json"
+
+  final val MaxRetries = 10
+  final val RetryBaseDelay = 1000 // milliseconds
+  final val RetryMaxDelay = 20 * 1000 // milliseconds
+
+  /** Common retry policy for S3 and SQS (jitter) */
+  final val RetryPolicy =
+    new RetryPolicy(
+      (_: AmazonWebServiceRequest, _: AmazonClientException, retriesAttempted: Int) =>
+        retriesAttempted < MaxRetries,
+      new PredefinedBackoffStrategies.FullJitterBackoffStrategy(RetryBaseDelay, RetryMaxDelay),
+      MaxRetries,
+      true
+    )
+
+  private final val MessageProcessorVersion = Semver
+    .decodeSemver(ProjectMetadata.version)
+    .fold(e => throw new IllegalStateException(s"Cannot parse project version $e"), identity)
+
+  final val MessageProcessor: LoaderMessage.Processor =
+    LoaderMessage.Processor(ProjectMetadata.shredderName, MessageProcessorVersion)
+
+
+  /** @return Tuple containing list of folders with incomplete shredding and list of unshredded folders */
+  def getState(region: String, enrichedFolder: Folder, shreddedFolder: Folder): (Set[S3.Folder], Set[S3.Folder]) = {
+    val client = createS3Client(region)
+    val enrichedDirs = Cloud.listDirs(client, enrichedFolder).toSet
+    val shreddedDirs = Cloud.listDirs(client, shreddedFolder).toSet
+
+    val enrichedFolderNames = enrichedDirs.map(Folder.coerce).map(_.folderName)
+    val shreddedFolderNames = shreddedDirs.map(Folder.coerce).map(_.folderName)
+
+    val incomplete = enrichedFolderNames.intersect(shreddedFolderNames).collect { 
+      case folder if !Cloud.keyExists(client, shreddedFolder.append(folder).withKey(FinalKeyName)) =>
+        enrichedFolder.append(folder)
+    }
+
+    val unshredded = enrichedFolderNames.diff(shreddedFolderNames)
+      .map(enrichedFolder.append)
+
+    (incomplete, unshredded)
+  }
+
+  def sendMessage(message: LoaderMessage.ShreddingComplete,
```

Comment:
It feels weird to have `sendMessage` in `Discovery`, should we move it to `ShredJob` instead ?

Reply:
Good point, I was also uncomfortable about it. I think `sendMessage` is just a bad name. Should be something like `checkpoint` (given that it also stores the success file on S3), `save`, `seal` or something alike - basically something that allows next run to discover the folder.

> should we move it to ShredJob instead ?

I did a big refactoring before leaving on holidays to unload the `ShredJob` module, it was doing way too much stuff unrelated to its name and it shrunk down from >600 LoC to 198, don't want to get back there.